include/exadg/fluid_structure_interaction/solver_precice.h

```diff
@@ -0,0 +1,221 @@
+                        dealii::Patterns::Integer(1,EXADG_DEGREE_MAX),
+                        true);
+      prm.add_parameter("DegreeStructure",
+                        degree_structure,
+                        "Polynomial degree of structural problem.",
+                        dealii::Patterns::Integer(1,EXADG_DEGREE_MAX),
+                        true);
+      prm.add_parameter("RefineFluid",
+                        refine_fluid,
+                        "Number of mesh refinements (fluid).",
+                        dealii::Patterns::Integer(0,20),
+                        true);
+      prm.add_parameter("RefineStructure",
+                        refine_structure,
+                        "Number of mesh refinements (structure).",
+                        dealii::Patterns::Integer(0,20),
+                        true);
+    prm.leave_subsection();
+    // clang-format on
+  }
+
+  unsigned int degree_fluid = 3, degree_structure = 3;
+
+  unsigned int refine_fluid = 0, refine_structure = 0;
+};
+
+
+void
+create_input_file(std::string const & input_file)
+{
+  dealii::ParameterHandler prm;
+
+  GeneralParameters general;
+  general.add_parameters(prm);
+
+  ResolutionParameters resolution;
+  resolution.add_parameters(prm);
+
+  ExaDG::preCICE::ConfigurationParameters precice_parameters;
+  precice_parameters.add_parameters(prm);
+
+  // we have to assume a default dimension and default Number type
+  // for the automatic generation of a default input file
+  constexpr int Dim = 2;
+  using Number      = double;
+
+  try
+  {
+    FSI::get_application<Dim, Number>(input_file, MPI_COMM_WORLD)->add_parameters(prm);
+  }
+  catch(...)
+  {
+  }
+
+  prm.print_parameters(input_file,
+                       dealii::ParameterHandler::Short |
+                         dealii::ParameterHandler::KeepDeclarationOrder);
+}
+
+
+template<int dim, typename Number>
+void
+run(std::string const & input_file, MPI_Comm const & mpi_comm, bool const is_test)
+{
+  dealii::Timer timer;
+  timer.restart();
+
+  std::shared_ptr<FSI::ApplicationBase<dim, Number>> application =
+    FSI::get_application<dim, Number>(input_file, mpi_comm);
+
+
+  ExaDG::ResolutionParameters resolution(input_file);
+  application->set_parameters_convergence_study(resolution.degree_fluid,
+                                                resolution.degree_structure,
+                                                resolution.refine_fluid,
+                                                resolution.refine_structure);
+
+  ExaDG::preCICE::ConfigurationParameters            precice_param(input_file);
+  std::shared_ptr<FSI::preCICE::Driver<dim, Number>> driver;
+
+  // We have different drivers for Fluid and Solid since the data exchange and underlying
+  // configuration are different. However, both solver share a common base class. One could also
+  // compile different executables for both solvers or try to create a common Driver class for both
+  // parts.
```

Comment:
I was thinking more about this comment. I think it would be easier to include `DriverSolid` and `DriverFluid` into a single class `preCICE::Driver`. I think the use of a base class + inheritance is unnecessarily complicated here. Instead, the class `preCICE::Driver` should use `WrapperFluid` and `WrapperStructure` proposed in PR #174.

The main reason for this conviction is the following: Currently, the fluid and structure problems are solved simultaneously, while some MPI ranks work on the fluid and some on the structure problem, right? However, I think it would add great value to the present implementation if one could switch to an iteration procedure that solves the fluid and structure problems sequentially, with all MPI ranks solving the fluid problem (and the structure problem), close to how the exadg-native FSI implementation does it. Regarding precice, I am referring to this publication: https://www.sciencedirect.com/science/article/pii/S0898122115005933 and the different coupling schemes described there. I think it would be easier to realize these different variants if only a single `Driver` class was used? Put differently, I think realizing other variants would again lead to code duplication.

I think another (minor) advantage would be that the if-else branch below would not be necessary, so that also the precice-FSI solver.h file would look like the other ExaDG solver.h files.

Reply:
The coupling scheme is independent of the current implementation. Each rank runs on his own MPI ranks and you can configure if you want them to run in a staggered way or simultaneously. If you run the simulation staggered, the compute cores of the non-computing participants are actually idling (which lead to the invention of parallel coupling schemes). Running two preCICE instances (two different participant) contradicts the partitioned approach to some extend (why do you use preCICE then and not exadg?). There was indeed recently some interest in such a case and it will work once https://github.com/precice/precice/pull/1190 is merged. 

Considering that this suggestion would require major changes once more I would prefer to keep it as an open TODO for the future rather than bothering now with it.